logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,169 @@
+
+            generator.writeFieldName("uri");
+            reconstructUri(request, generator);
+            
+            writeHeaders(generator, request);
+    
+            writeBody(request, body, generator);
+    
+            generator.writeEndObject();
         }
-        if (JsonMediaType.JSON.test(contentType)) {
-        	// TODO has this JSON been validated? If not then this might result in invalid log statements
-            return Optional.of(new JsonBody(body));
-        } else {
-            return Optional.of(body);
+        
+        return writer.toString();
+            
+    }
+
+    private void writeBody(HttpMessage message, String body, JsonGenerator builder) throws IOException {
+        if(!body.isEmpty()) {
+            builder.writeFieldName("body");
+
+            final String contentType = message.getContentType();
+
+            if (JsonMediaType.JSON.test(contentType)) {
+                builder.writeRawValue(body);
+            } else {
+                builder.writeString(body);
+            }
         }
     }
 
     @Override
-    public String format(final Map<String, Object> content) throws IOException {
-        return mapper.writeValueAsString(content);
+    public String format(Correlation correlation, HttpResponse response) throws IOException {
+        
+        final String correlationId = correlation.getId();
+        
+        String body = response.getBodyAsString();
+
+        StringWriter writer = new StringWriter(body.length() + 2048);
+
+        try (JsonGenerator generator = jsonFactory.createGenerator(writer)) {
+            generator.writeStartObject();
+    
+            generator.writeFieldName("origin");
+            if(response.getOrigin() == Origin.LOCAL) {
+                generator.writeRawValue("\"local\"");
+            } else {
+                generator.writeRawValue("\"remote\"");
+            }
+            generator.writeFieldName("type");
+            generator.writeRawValue("\"response\"");
+            generator.writeStringField("correlation", correlationId);
+            generator.writeStringField("protocol", response.getProtocolVersion());
+            generator.writeNumberField("duration", correlation.getDuration().toMillis());
+            generator.writeNumberField("status", response.getStatus());
+    
+            writeHeaders(generator, response);
+            writeBody(response, body, generator);
+            
+            generator.writeEndObject();
+        }
+        return writer.toString();
     }
 
-    @AllArgsConstructor
-    private static final class JsonBody {
-        String json;
+    protected void writeHeaders(JsonGenerator builder, HttpMessage httpMessage) throws IOException {
+        Map<String, List<String>> headers = httpMessage.getHeaders();
+        
+        if(!headers.isEmpty()) {
+            // implementation note:
+            // for some unclear reason, manually iterating over the headers
```

Comment:
If that is the case for headers, maybe sticking to a map-based approach (see `StructuredHttpLogFormatter`) would yield similar results without introducing a breaking API change?

Reply:
Not 100% sure, but I think it has something to do with a smart way of serializing arrays/Lists in a Map. I'm able to make something thats almost as fast. I'm sure there is no magic, but did not pursue it further due to time, so could be a future improvement.

But to answer your question, no, then the performance improvement would be gone?